schema/encoding/jsonschema.go

```diff
@@ -0,0 +1,165 @@
+func validatorToJSONSchema(w io.Writer, v schema.FieldValidator) (err error) {
+	if v == nil {
+		return nil
+	}
+	ew := errWriter{w: w}
+	switch t := v.(type) {
+	case *schema.String:
+		ew.writeString(`"type": "string"`)
+		if t.Regexp != "" {
+			ew.writeFormat(`, "pattern": %q`, t.Regexp)
+		}
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, fmt.Sprintf("%q", value))
+			}
+			ew.writeFormat(`, "enum": [%s]`, strings.Join(allowed, ", "))
+		}
+		if t.MinLen > 0 {
+			ew.writeFormat(`, "minLength": %s`, strconv.FormatInt(int64(t.MinLen), 10))
+		}
+		if t.MaxLen > 0 {
+			ew.writeFormat(`, "maxLength": %s`, strconv.FormatInt(int64(t.MaxLen), 10))
+		}
+	case *schema.Integer:
+		ew.writeString(`"type": "integer"`)
+
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, strconv.FormatInt(int64(value), 10))
+			}
+			ew.writeFormat(`, "enum": [%s]`, strings.Join(allowed, ","))
+		}
+		if t.Boundaries != nil {
+			ew.writeFormat(`, "minimum": %s, "maximum": %s`,
+				strconv.FormatFloat(t.Boundaries.Min, 'E', -1, 64),
+				strconv.FormatFloat(t.Boundaries.Max, 'E', -1, 64))
+		}
+	case *schema.Float:
+		ew.writeString(`"type": "number"`)
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, strconv.FormatFloat(value, 'E', -1, 64))
+			}
+			ew.writeFormat(`, "enum": [%s]`, strings.Join(allowed, ","))
+		}
+		if t.Boundaries != nil {
+			ew.writeFormat(`, "minimum": %s, "maximum": %s`,
+				strconv.FormatFloat(t.Boundaries.Min, 'E', -1, 64),
+				strconv.FormatFloat(t.Boundaries.Max, 'E', -1, 64))
+		}
+
+	case *schema.Array:
+		ew.writeString(`"type": "array"`)
+		if t.ValuesValidator != nil {
+			ew.writeString(`, "items": `)
+			if ew.err == nil {
+				ew.err = validatorToJSONSchema(w, t.ValuesValidator)
+			}
+		}
+	case *schema.Object:
+		if ew.err == nil {
+			ew.err = schemaToJSONSchema(w, t.Schema)
+		}
+	case *schema.Time:
+		ew.writeString(`"type": "string", "format": "date-time"`)
+	case *schema.Bool:
+		ew.writeString(`"type": "boolean"`)
+	default:
+		return ErrNotImplemented
+	}
+	return ew.err
+}
+
+// SchemaToJSONSchema helper
+func schemaToJSONSchema(w io.Writer, s *schema.Schema) (err error) {
+	ew := errWriter{w: w}
+	ew.writeString("{")
+	if s.Description != "" {
+		ew.writeFormat(`"title": %q, `, s.Description)
+	}
+	ew.writeString(`"type": "object", `)
+	ew.writeString(`"properties": {`)
+	var required []string
+	var notFirst bool
+	for key, field := range s.Fields {
+		if notFirst {
+			ew.writeString(", ")
+		}
+		notFirst = true
+		ew.writeFormat("%q: {", key)
+		if field.Description != "" {
+			ew.writeFormat(`"description": %q, `, field.Description)
+		}
+		if field.Required {
+			required = append(required, fmt.Sprintf("%q", key))
+		}
+		if field.ReadOnly {
+			ew.writeFormat(`"readOnly": %t, `, field.ReadOnly)
+		}
```

Comment:
I've avoided default handling for the moment because it can be a function (onInit()) or a fixed value interface{} value. Did you mean %T? That would also use the go type, which might not be quite right. What do you think we should do here?